### ISSUE TITLE
Feature request: 32bit support

### ISSUE THREAD

**Issue by IamTheCarl** [khvzak/mlua#27]
I'm in a unique situation where I need to compile Lua source to Lua 32bit binary code to send off to an embedded device. I've proven the technique works but it requires a 32bit Lua implementation on the host machine as well.

To fix this requires simply adding the C define LUA_32BITS from the lua-src crate. From what I understand, I think we would have to pass the feature flag to enable that from this crate.

Is it doable?

**Comment by khvzak** [comment]
As an option, you could compile your own Lua with `-DLUA_32BITS` and use it together with mlua.
I've tested it on my laptop:
```
$ LUA_INC=$HOME/tmp/lua-5.4.2-32bit/src LUA_LIB=$HOME/tmp/lua-5.4.2-32bit/src LUA_LIB_NAME=lua LUA_LINK=static cargo build

$ cat ./target/debug/build/mlua-e5620f4236dc7a6a/out/glue.rs | rg 'LUA_NUMBER|LUA_INTEGER|LUA_UNSIGNED'
pub type LUA_NUMBER = c_float;
pub type LUA_INTEGER = i32;
pub type LUA_UNSIGNED = u32;
```

The option to provide `LUA_INC`/`LUA_LIB`/etc is designed especially for such cases, to work with custom Lua libraries.
Or do you rely on the `vendored` feature and/or cross-compilation?

**Comment by IamTheCarl** [comment]
Yes, that's how I proved this works. It just feels inconvient to add that extra step to the build command. I really admire Cargo's plug and play expirence with building, and I'd like to keep true to that.

After sleeping on this I came up with a simpler solution. What if we added a feature flag to disable both the vendored version and the pkg-config? It's then up to the user to compile/link in Lua themselves using something like the CC crate from the build.rs. This way they get the option of a customized Lua, and can keep the convinence of not having to provide environment variables.

It also unlocks the potential for third party vendored crates, since they could link in Lua themselves as well.

**Comment by khvzak** [comment]
>  It's then up to the user to compile/link in Lua themselves using something like the CC crate from the build.rs. 
I believe this feature is already provided out of the box by lovely Cargo.

I added an example how to use a customised Lua version with mlua that works by simply running `cargo build`.
https://github.com/khvzak/mlua-usecases/tree/main/lua-32bit
It uses Lua 5.3.6 with uncommented [`#define LUA_32BITS`](https://github.com/khvzak/mlua-usecases/blob/main/lua-32bit/lua-src/lua-5.3.6/luaconf.h#L36).

**Comment by IamTheCarl** [comment]
Oh of course!
I tried doing this but the vendored feature kept linking in the default lua-src, causing me problems.
I didn't think to just patch in my own version. I'll try that when I can.

**Comment by IamTheCarl** [comment]
For the most part this has worked but there is a small problem I should note.

I was making use of the serialization feature. It expected 64bit types. It's now getting 32bit types. Because of this that code doesn't compile anymore.

I was planning to customize the serialization so I could send tables of functions to the embedded device anyway, so this doesn't really set me back much. I just disabled it for now.

Aside from that, I can now send a pre-compiled function to the device and run it without an issue. Thank you very much.

**Comment by khvzak** [comment]
@IamTheCarl thanks for the bug report with serialisation, I'll fix that.

**Comment by khvzak** [comment]
I've released 0.5.1 that should have fixed serialisation for 32bit Lua.
